feat(options): add 'eventignorewin'

luukvbaal opened the pull request:
**vim-patch:partial:9.1.1084: Unable to persistently ignore events in a window and its buffers**

Problem:  Unable to persistently ignore events in a window and its buffers.
Solution: Add 'eventignorewin' option to ignore events in a window and buffer
          (Luuk van Baal)

Add the window-local 'eventignorewin' option that is analogous to
'eventignore', but applies to a certain window and its buffers. Identify
events that should be allowed in 'eventignorewin', adapt "auto_event"
and "event_tab" to encode this information. Window context is not passed
onto apply_autocmds_group(), and when to ignore an event is a bit
ambiguous when "buf" is not "curbuf", rather than a large refactor, only
ignore an event when all windows into "buf" are ignoring the event.

https://github.com/vim/vim/commit/b7147f8236c962cd74d1ce028d9106f1c449ea6c

**vim-patch:9.1.1102: tests: Test_WinScrolled_Resized_eiw() uses wrong filename**

Problem:  tests: Test_WinScrolled_Resized_eiw() uses wrong filename
          (Luuk van Baal, after v9.1.1084)
Solution: Rename the filename to something more unique

https://github.com/vim/vim/commit/bfc7719e48ffc365ee0a1bd1888120d26b6365f0

Resolve #32055

lewis6991 commented:
Does this lock us out ever making this buffer local in the future?

luukvbaal commented:
> Does this lock us out ever making this buffer local in the future?

Without #29314, yes. Would buffer local be better do you think? There are both window and buffer autocommands so both would come with caveats.

lewis6991 commented:
I have no idea tbh. Buffer local is sort of more granular and easier to reason about, so seems like a slightly safer option. With that said, a large majority of global-local options are window local.

Maybe worth asking around vim/vim too?

luukvbaal commented:
> Buffer local is sort of more granular and easier to reason about, so seems like a slightly safer option.

I think so too, but it does seem like it would be a bit of an outlier when compared to current buffer-local options. Maybe justified considering buffer events though, and simple enough to reason about ignoring window events that contain a buffer with 'eventignore' set.

przepompownia commented:
Thanks for starting to try to solve the problem! I started to see how it behaves even at this early stage.

```lua
local api = vim.api
local caseInsensivity = nil

api.nvim_create_autocmd('CmdLineEnter', {
  group = api.nvim_create_augroup('cmdlineIgnoreCase', {clear = true}),
  pattern = ':',
  callback = function ()
    caseInsensivity = vim.go.ignorecase
    vim.opt.ignorecase = true
  end
})
api.nvim_create_autocmd('CmdLineLeave', {
  group = augroup,
  pattern = ':',
  callback = function ()
    vim.go.ignorecase = caseInsensivity
    caseInsensivity = nil
  end
})

local msgBuf = api.nvim_create_buf(false, true)
local msgWin = api.nvim_open_win(msgBuf, false, {
  relative = 'editor',
  row = vim.go.lines - 2,
  col = vim.o.columns,
  width = 60,
  height = 1,
  anchor = 'SE',
  style = 'minimal',
  focusable = false,
  zindex = 999,
  noautocmd = true,
})
vim.wo[msgWin].eventignore = 'all'

api.nvim_win_set_config(msgWin, {
  title = tostring(msgWin),
  border = 'single',
})
```

With this init file cmdline completion result of `:i` does not contain `:Inspect` (unlike without `vim.wo[msgWin].eventignore = 'all'`). It suggests that (confirmed after `:se verbose=9` on both cases) `CmdLineEnter` autocommands are not executed. Is it correct for autocommands not related to window (assuming the current state of this PR)? I can try to catch similar behavior on some `Win*` events.

przepompownia commented:
```lua
local api = vim.api
vim.wo.number = true

local function fill()
  api.nvim_buf_set_lines(1, -1, -1, false, {tostring(vim.uv.hrtime()/1000000)})
end

for i = 1, vim.o.lines + 1 do
  fill()
end

api.nvim_create_autocmd('WinScrolled', {
  group = api.nvim_create_augroup('', {clear = true}),
  pattern = '*',
  callback = fill
})

local msgWin = api.nvim_open_win(api.nvim_create_buf(false, true), false, {
  relative = 'editor',
  row = vim.go.lines - 2,
  col = vim.o.columns,
  width = 60,
  height = 1,
  anchor = 'SE',
  style = 'minimal',
  focusable = false,
  zindex = 999,
  noautocmd = true,
})

vim.wo[msgWin].eventignore = 'WinScrolled'
```

`vim.wo[msgWin].eventignore = 'WinScrolled'` prevents from executing `WinScrolled` not only on `msgWin`.

luukvbaal commented:
> Is it correct for autocommands not related to window

No that shouldn't happen, in which case we should maybe restrict the allowed values to buffer/window related autocommands only for the local value.

> vim.wo[msgWin].eventignore = 'WinScrolled' prevents from executing WinScrolled not only on msgWin.

PR is misusing the `b_wininfo` list, need to loop over all windows instead.

luukvbaal commented:
WIP at https://github.com/vim/vim/pull/16530.

luukvbaal committed:
feat(options): add 'eventignorewin'

vim-patch:partial:9.1.1084: Unable to persistently ignore events in a window and its buffers

Problem:  Unable to persistently ignore events in a window and its buffers.
Solution: Add 'eventignorewin' option to ignore events in a window and buffer
          (Luuk van Baal)

Add the window-local 'eventignorewin' option that is analogous to
'eventignore', but applies to a certain window and its buffers. Identify
events that should be allowed in 'eventignorewin', adapt "auto_event"
and "event_tab" to encode this information. Window context is not passed
onto apply_autocmds_group(), and when to ignore an event is a bit
ambiguous when "buf" is not "curbuf", rather than a large refactor, only
ignore an event when all windows into "buf" are ignoring the event.

https://github.com/vim/vim/commit/b7147f8236c962cd74d1ce028d9106f1c449ea6c

vim-patch:9.1.1102: tests: Test_WinScrolled_Resized_eiw() uses wrong filename

Problem:  tests: Test_WinScrolled_Resized_eiw() uses wrong filename
          (Luuk van Baal, after v9.1.1084)
Solution: Rename the filename to something more unique

https://github.com/vim/vim/commit/bfc7719e48ffc365ee0a1bd1888120d26b6365f0

luukvbaal commented:
This now (partially) ports https://github.com/vim/vim/commit/b7147f82. Partial port because of conflicts with https://github.com/vim/vim/commit/1755a918 and https://github.com/vim/vim/commit/78d742ab, which are as of yet unported. We could port those first but I doubt we would want to. The former adds `autocmd_get()` for which we have `nvim_get_autocmds()`, the latter adds a keyvalue table for an autocmd list which we already generate.

przepompownia commented:
It seems to just work for me at first glance.

zeertzjq commented on test/functional/legacy/autocmd_spec.lua:
```diff
@@ -40,3 +41,60 @@ it('no E440 in quickfix window when autocommand invalidates undo', function()
+-- oldtest: Test_WinScrolled_Resized_eiw()
+it('WinScrolled and WinResized events can be ignored in a window', function()
+  local screen = Screen.new()
+  write_file(
+    'Xtest_winscrolled_mouse',
+    [[
+    call setline(1, ['foo']->repeat(32))
+    set eventignorewin=WinScrolled,WinResized
+    split
+    let [g:afile,g:resized,g:scrolled] = ['none',0,0]
+    au WinScrolled * let [g:afile,g:scrolled] = [expand('<afile>'),g:scrolled+1]
+    au WinResized * let [g:afile,g:resized] = [expand('<afile>'),g:resized+1]
+    ]]
+  )
+  finally(function()
+    os.remove('Xtest_winscrolled_mouse')
+  end)
+  command('so Xtest_winscrolled_mouse')
```
Since you don't need to edit the file, you can just use `exec()` without writing the script to a file:
```suggestion
  n.exec([[
    call setline(1, ['foo']->repeat(32))
    set eventignorewin=WinScrolled,WinResized
    split
    let [g:afile,g:resized,g:scrolled] = ['none',0,0]
    au WinScrolled * let [g:afile,g:scrolled] = [expand('<afile>'),g:scrolled+1]
    au WinResized * let [g:afile,g:resized] = [expand('<afile>'),g:resized+1]
  ]])
```

zeertzjq commented on src/nvim/auevents.lua:
```diff
@@ -1,156 +1,139 @@
 return {
+  --- @type [string, string[], boolean][] List of [eventname, aliases, window-local event] tuples.
   events = {
```
This annotation suggests that an event can have multiple aliases, but both `gen_events.lua` and `gen_vimvim.lua` only support a single alias per event. Perhaps you should add `ipairs()` to support multiple aliases?

zeertzjq commented on src/nvim/generators/gen_events.lua:
```diff
@@ -16,46 +15,28 @@ names_tgt:write([[
-
-names_tgt:write('\n  {0, NULL, (event_T)0},')
+names_tgt:write('\n  {0, NULL, (event_T)0},\n};\n')
+names_tgt:write('\nstatic AutoCmdVec autocmds[NUM_EVENTS] = { KV_INITIAL_VALUE };\n')
```
This doesn't make sense. It only initializes the first element of `autocmds[]` to `KV_INITIAL_VALUE`, although `KV_INITIAL_VALUE` is all zero anyway...

zeertzjq replied:
Hmm, or just change it to `0`?

justinmk commented:
> conflicts with [vim/vim@1755a918](https://github.com/vim/vim/commit/1755a918) and [vim/vim@78d742ab](https://github.com/vim/vim/commit/78d742ab), which are as of yet unported. We could port those first but I doubt we would want to

Yeah we don't want those. Please add them as N/A patches to the commit message in this PR. And optionally mention those functions in `:help nvim-removed`, though we usually don't list all functions there (but maybe we should).

zeertzjq commented:
I don't think the patches should be marked as N/A, because there are conflicts in the tests.
